jlibra-core/src/main/java/dev/jlibra/mnemonic/ExtendedPrivKey.java

```diff
@@ -1,72 +1,43 @@
 
 import javax.annotation.concurrent.Immutable;
-import java.nio.ByteBuffer;
+
+import org.bouncycastle.asn1.edec.EdECObjectIdentifiers;
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters;
+import org.bouncycastle.crypto.util.PrivateKeyInfoFactory;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+
+import dev.jlibra.KeyUtils;
 
 @Immutable
 public class ExtendedPrivKey {
 
-    public final SecretKey privateKey;
-    public final SecretKey publicKey;
+    public final PrivateKey privateKey;
+    public final PublicKey publicKey;
 
     public ExtendedPrivKey(SecretKey secretKey) {
-        this.privateKey = secretKey;
-        this.publicKey = new SecretKey(createPublic(secretKey));
-    }
-
-    private static byte[] createPublic(SecretKey secretKey) {
-        return new Ed25519PrivateKeyParameters(secretKey.getData(), 0)
-                .generatePublicKey()
-                .getEncoded();
-    }
-
-    public String getAddress() {
-        SHA3.Digest256 digest256 = new SHA3.Digest256();
-        digest256.update(publicKey.getData());
-        return Hex.toHexString(digest256.digest());
-    }
-
-    public byte[] sign(RawTransaction rawTransaction) {
-        SHA3.DigestSHA3 digestSHA3 = new SHA3.Digest256();
-        byte[] saltDigest = digestSHA3.digest("RawTransaction@@$$LIBRA$$@@".getBytes());
-        byte[] transactionBytes = rawTransaction.toByteArray();
-        byte[] messageBytes = ByteBuffer.allocate(saltDigest.length + transactionBytes.length)
-                .put(saltDigest)
-                .put(transactionBytes)
-                .array();
-        byte[] message = digestSHA3.digest(messageBytes);
-
-        return sign(message);
-    }
+        Ed25519PrivateKeyParameters pKeyParams = new Ed25519PrivateKeyParameters(secretKey.getData(), 0);
 
-    @VisibleForTesting
-    byte[] sign(byte[] message) {
         try {
-            Ed25519Signer signer = new Ed25519Signer();
-            signer.init(true, new Ed25519PrivateKeyParameters(privateKey.getData(), 0));
-            signer.update(message, 0, message.length);
-            return signer.generateSignature();
-        } catch (Exception e) {
-            throw new RuntimeException("Signing the transaction failed", e);
+            PrivateKeyInfo keyInfo = PrivateKeyInfoFactory.createPrivateKeyInfo(pKeyParams);
+            this.privateKey = KeyUtils.getKeyFactory().generatePrivate(new PKCS8EncodedKeySpec(keyInfo.getEncoded()));
+            this.publicKey = BouncyCastleProvider.getPublicKey(new SubjectPublicKeyInfo(
+                    new AlgorithmIdentifier(EdECObjectIdentifiers.id_Ed25519), keyInfo.getPublicKeyData().getBytes()));
+        } catch (IOException | InvalidKeySpecException e) {
+            throw new RuntimeException("Key creation failed", e);
```

Comment:
Nice, that's what I couldn't find out. If this works, I could also restructure the wallet in the cli which currently separates between "generic" accounts and mnemonic accounts. Without this distinction, I would be way easier to handle wallets.

Reply:
this is great